pkg/reconciler/tenancy/clusterworkspace/clusterworkspace_reconcile_scheduling.go

```diff
@@ -33,16 +40,77 @@ import (
 	"github.com/kcp-dev/kcp/pkg/apis/third_party/conditions/util/conditions"
 )
 
+const (
+	// thisWorkspaceFinalizer attached to new ClusterWorkspace (in phase ClusterWorkspacePhaseScheduling) resources so that we can control
+	// deletion of ThisWorkspace resources
+	thisWorkspaceFinalizer = "tenancy.kcp.dev/thisworkspace"
+
+	// clusterWorkspaceShardAnnotationKey keeps track on which shard ThisWorkspace must be scheduled
+	clusterWorkspaceShardAnnotationKey = "tenancy.kcp.dev/shard"
+)
+
 type schedulingReconciler struct {
 	getShard   func(name string) (*tenancyv1alpha1.ClusterWorkspaceShard, error)
 	listShards func(selector labels.Selector) ([]*tenancyv1alpha1.ClusterWorkspaceShard, error)
 }
 
 func (r *schedulingReconciler) reconcile(ctx context.Context, workspace *tenancyv1alpha1.ClusterWorkspace) (reconcileStatus, error) {
 	logger := klog.FromContext(ctx)
-	workspaceClusterName := logicalcluster.From(workspace)
 	switch workspace.Status.Phase {
 	case tenancyv1alpha1.ClusterWorkspacePhaseScheduling:
+		// TODO (p0lyn0mial): we could store a hash over a shard name and then use an index to find a real value
+		shardName, hasShardName := workspace.Annotations[clusterWorkspaceShardAnnotationKey]
+		if !hasShardName {
+			// note that the annotation can be only removed by a system:masters, so we should be fine here.
+			var err error
+			shardName, err = r.chooseShardAndMarkCondition(logger, workspace)
+			if err != nil {
+				return reconcileStatusStopAndRequeue, err
+			}
+			if len(shardName) == 0 {
+				return reconcileStatusContinue, nil // retry is automatic when new shards show up
+			}
+			addShardAnnotation(workspace, shardName)
+			// this is the first part of our two-phase commit
+			// the first phase is to pick up a shard
+			return reconcileStatusContinue, nil
+		}
+
+		shard, err := r.getShard(shardName)
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				conditions.MarkFalse(workspace, tenancyv1alpha1.WorkspaceScheduled, tenancyv1alpha1.WorkspaceReasonUnschedulable, conditionsv1alpha1.ConditionSeverityError, "chosen shard %q does not exist anymore: %v", shardName, err)
+				return reconcileStatusContinue, nil
+			}
+			return reconcileStatusStopAndRequeue, err
+		}
+		if valid, reason, message := isValidShard(shard); !valid {
+			conditions.MarkFalse(workspace, tenancyv1alpha1.WorkspaceScheduled, tenancyv1alpha1.WorkspaceReasonUnschedulable, conditionsv1alpha1.ConditionSeverityError, "chosen shards %q is no longer valid, reason %q, message %q", shardName, message, reason)
+			return reconcileStatusStopAndRequeue, nil
+		}
+
+		if !hasThisWorkspaceFinalizer(workspace) {
+			workspace.Finalizers = append(workspace.Finalizers, thisWorkspaceFinalizer)
```

Comment:
this likely means we have to return here because this is technically a spec change.

To avoid too many writes, move this to `addShardAnnotation`? I did that in the doc.

Reply:
yes, wanted to do the same https://github.com/kcp-dev/kcp/pull/2327#discussion_r1019113615
anyway i think we need the check in both places - just in case.